diff -Naur Chapter4_1/Cpu0InstrInfo.td Chapter4_2/Cpu0InstrInfo.td
--- Chapter4_1/Cpu0InstrInfo.td	2013-07-05 17:46:02.913283807 +0800
+++ Chapter4_2/Cpu0InstrInfo.td	2013-07-05 17:46:02.905283807 +0800
@@ -230,3 +230,14 @@
 
 def : Pat<(i32 immSExt16:$in),
           (ADDiu ZERO, imm:$in)>;
+
+def : Pat<(not CPURegs:$in),
+          (XOR CPURegs:$in, (ADDiu ZERO, 1))>;
+
+// setcc patterns
+multiclass SeteqPats<RegisterClass RC, Instruction XOROp> {
+  def : Pat<(seteq RC:$lhs, RC:$rhs),
+            (XOROp (XOROp RC:$lhs, RC:$rhs), (ADDiu ZERO, 1))>;
+}
+
+defm : SeteqPats<CPURegs, XOR>;
