diff -Naur Chapter8_4/Cpu0FrameLowering.cpp Chapter8_5/Cpu0FrameLowering.cpp
--- Chapter8_4/Cpu0FrameLowering.cpp	2013-06-26 12:36:34.777595199 +0800
+++ Chapter8_5/Cpu0FrameLowering.cpp	2013-06-26 12:36:34.777595199 +0800
@@ -212,6 +212,15 @@
     assert((StackSize <= 0x7fffffffULL) && "Stack Size out of range! StackSize must <= 0x7fffffff)");
 }
 
+// This function eliminate ADJCALLSTACKDOWN,
+// ADJCALLSTACKUP pseudo instructions
+void Cpu0FrameLowering::
+eliminateCallFramePseudoInstr(MachineFunction &MF, MachineBasicBlock &MBB,
+                              MachineBasicBlock::iterator I) const {
+  // Simply discard ADJCALLSTACKDOWN, ADJCALLSTACKUP instructions.
+  MBB.erase(I);
+}
+
 // This method is called immediately before PrologEpilogInserter scans the 
 //  physical registers used to determine what callee saved registers should be 
 //  spilled. This method is optional. 
diff -Naur Chapter8_4/Cpu0FrameLowering.h Chapter8_5/Cpu0FrameLowering.h
--- Chapter8_4/Cpu0FrameLowering.h	2013-06-26 12:36:34.777595199 +0800
+++ Chapter8_5/Cpu0FrameLowering.h	2013-06-26 12:36:34.777595199 +0800
@@ -35,6 +35,9 @@
   void emitPrologue(MachineFunction &MF) const;
   void emitEpilogue(MachineFunction &MF, MachineBasicBlock &MBB) const;
   bool hasFP(const MachineFunction &MF) const;
+  void eliminateCallFramePseudoInstr(MachineFunction &MF,
+                                     MachineBasicBlock &MBB,
+                                     MachineBasicBlock::iterator I) const;
   void processFunctionBeforeCalleeSavedScan(MachineFunction &MF,
                                             RegScavenger *RS) const;
 };
diff -Naur Chapter8_4/Cpu0InstrInfo.cpp Chapter8_5/Cpu0InstrInfo.cpp
--- Chapter8_4/Cpu0InstrInfo.cpp	2013-06-26 12:36:34.777595199 +0800
+++ Chapter8_5/Cpu0InstrInfo.cpp	2013-06-26 12:36:34.777595199 +0800
@@ -21,7 +21,7 @@
 using namespace llvm;
 
 Cpu0InstrInfo::Cpu0InstrInfo(Cpu0TargetMachine &tm)
-  : 
+  : Cpu0GenInstrInfo(Cpu0::ADJCALLSTACKDOWN, Cpu0::ADJCALLSTACKUP),
     TM(tm),
     RI(*TM.getSubtargetImpl(), *this) {}
 
